### PR TITLE
CLOUDP-114735: decrypt logs via Azure KeyVault

### PR DESCRIPTION
**Pull request by fmenezes**
## Proposed changes
decrypt logs via Azure KeyVault

_Jira ticket:_ CLOUDP-114735

Closes #[issue number]

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongocli/blob/master/CONTRIBUTING.md) (if appropriate)
- [x] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](https://github.com/mongodb/mongocli/blob/master/e2e/E2E-TESTS.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

My steps to test this
1. Create symmetric key in Azure
  - I basically followed https://docs.microsoft.com/en-us/azure/key-vault/keys/quick-create-portal
2. To prepare test data I've started with a local key file
- `loglines.json`
```json
{"ts":{"$date":{"$numberLong":"1644232049921"}},"version":"0.0","compressionMode":"zstd","keyStoreIdentifier":{"provider":"local","filename":"localKey"},"encryptedKey":{"$binary":{"base64":"+yjPCaKKE1M8fZmPGzGHkyfHYxaw34okpavsHzpd8iPVx2+JjOhXwXw5E2FdI5Rcb5JgmcPUFRPISh/7Si1R/g==","subType":"0"}},"MAC":"qE9fUsGK0EuRrrCRAQAAAAAAAAAAAAAA","auditRecordType":"header"}
{"ts":{"$date":{"$numberLong":"1644232049922"}},"log":"1Lu4o8XVMM/Rg7GKAQAAAAEAAAAAAAAA/8tXQ36mEd90OaAOzCOSti7N5a2jr0B9ek48/uvyteG/zUJHyM16Hs3wMEhDqTQGBwGhWSHEqXh0/5Jbz6tXsYHhDTMr1BOsn1zaavZScx/CkO5+Hd8Vx+zeFPREtQTe1y+JngXSIroezeyV0/zF4YC4vpug+OZtrEQLNEgwT2bjaqUyaKDbmzCNetd2Ff/eFfMFzinbzKVgXAC7T4YmDuowqXommEXLIBiYh2u4VagwJKZRw5OGZjnvqwyVpSPgGqLxGKUoFigh3NgC6EuGi17VIs5BLRZOIw7+OfbPgQQiKzjCxCk="}
{"ts":{"$date":{"$numberLong":"1644232050645"}},"log":"gjn0BYabxQku84PTAQAAAAIAAAAAAAAAuTcMKIEGtlkPyGGqcj3mVOW0d+SUKGREXwicsBC5oDs+f4GkiNQmwXzpsassKTGLCS30xoEe/2NK/pBB7hB0mUmmoEVXCqgT28aztDXlkN5GeX0ECCYG4HoQyzyVdIi/SFF1Xa7jTLkYGdqwghDaMD/PzJVXcdhCckqPUn9CZKurMnsVZ7KiL8hpgNpQ8Q=="}
{"ts":{"$date":{"$numberLong":"1644232050645"}},"log":"8nCVKIkN1ebPXTIFAQAAAAMAAAAAAAAAMJ6H+8m76AZ0nOAHAWUmd7iZVHepDNpWyIBhTo4C5R7gBHFsPOlr4IETL/v6I+HSU4KBNnOFpoCA7KurmwNwQ7J8xdjrteCWB7WDIzP9cusu+4lsBdkhL/ExIJAaL/Rn/8Kq9482DyPtlxOldSUBP8QlyEJQ/iZ5W2fpgKdZtjCD4g+r6NPOgGjyrtr5bSqWts2f9Omth3r073HViNw="}
```
- `localKey`
```
LxbM6ik6lgJjKEkugTdkRUoHCeyArNxg2xx7kGHl/io=
```
3. I've followed the steps to grab a local key decrypted
- Local Key decrypted (`lek.txt`)
```
pnvb++3sbhxIJdfODOq5uIaUX8yxTuWS95VLgES30FM=
```
4. I've converted from base64 to binary
```
base64 -d lek.txt > lek.bin
```
5. finally re-encrypted the key into key.encrypted.txt
```
az keyvault key encrypt --value $(cat lek.bin) --algorithm RSA-OAEP --vault-name <<vault name>> --name <<key name>> --subscription <<subscription_id>> | jq -rM '.result' > lek.encrypted.txt
```
6. Replaced header line (first line) in `loglines.json` with:
```json
{"ts":{"$date":{"$numberLong":"1644232049921"}},"version":"0.0","compressionMode":"zstd","keyStoreIdentifier":{"provider":"azure","keyname":"<<key name>>","keyvaultendpoint":"https://<<vault name>>.vault.azure.net","keyversion":"<<key version>>"},"encryptedKey":{"$binary":{"base64":"<<contents of lek.encrypted.txt>>","subType":"0"}},"MAC":"qE9fUsGK0EuRrrCRAQAAAAAAAAAAAAAA","auditRecordType":"header"}
```
7. ran decryption
```
go run ./cmd/atlas/atlas.go logs decrypt --file ./loglines.json --azureClientId <clientId> --azureTenantId <tenantId> --azureSecret <secret>
```

### CODE REVIEW

File: internal/decryption/keyproviders/azure.go [hunk 1]
```diff
@@ -83,6 +84,14 @@ func (ki *AzureKeyIdentifier) ValidateCredentials() error {
...
 }
 
-func (ki *AzureKeyIdentifier) DecryptKey(_ []byte) ([]byte, error) {
-	return nil, errors.New("not implemented")
+func (ki *AzureKeyIdentifier) DecryptKey(key []byte) ([]byte, error) {
+	client, err := crypto.NewClient(ki.KeyVaultEndpoint+"/keys/"+ki.KeyName+"/"+ki.KeyVersion, ki.credentials, nil)
+	if err != nil {
+		return nil, err
+	}
+	r, err := client.Decrypt(context.Background(), crypto.EncryptionAlgRSAOAEP, key, &crypto.DecryptOptions{})
```
Comment (fmenezes) [inline internal/decryption/keyproviders/azure.go:92]: the algorithm is hardcoded for now, we should check it again with server/atlas teams when this is implemented

File: internal/decryption/encrypted_audit_log.go [hunk 2]
```diff
@@ -27,26 +27,26 @@ type AuditRecordType string
 type AuditLogLineKeyStoreIdentifier struct {
 	Provider *keyproviders.KeyStoreProvider `json:"provider,omitempty"`
 	// localKey
-	Filename *string `json:"filename,omitempty"`
+	Filename string `json:"filename,omitempty"`
```
Comment (gssbzn) [inline internal/decryption/encrypted_audit_log.go:30]: 👀  made sure empty string is not valid or remove `omitempty`

File: internal/decryption/encrypted_audit_log.go [hunk 3]
```diff
@@ -27,26 +27,26 @@ type AuditRecordType string
...
 	// kmip
-	UID            *string                         `json:"uniqueKeyID,omitempty"`
-	KMIPServerName []string                        `json:"kmipServerName,omitempty"`
-	KMIPPort       *int                            `json:"kmipPort,omitempty"`
-	KeyWrapMethod  *keyproviders.KMIPKeyWrapMethod `json:"keyWrapMethod,omitempty"`
+	UID            string                         `json:"uniqueKeyID,omitempty"`
```
Comment (gssbzn) [inline internal/decryption/encrypted_audit_log.go:32]: same for all these pointer removals, be sure zero or empty strings are not valid values or remove the `omitempty`

File: internal/decryption/keyproviders/azure.go (hunk 1 above)
Comment (gssbzn) [reply internal/decryption/keyproviders/azure.go:92]: link to a ticket for the follow up please

File: internal/decryption/keyproviders/azure.go [hunk 4]
```diff
@@ -83,6 +84,14 @@ func (ki *AzureKeyIdentifier) ValidateCredentials() error {
 	return err
 }
 
-func (ki *AzureKeyIdentifier) DecryptKey(_ []byte) ([]byte, error) {
-	return nil, errors.New("not implemented")
+func (ki *AzureKeyIdentifier) DecryptKey(key []byte) ([]byte, error) {
+	client, err := crypto.NewClient(ki.KeyVaultEndpoint+"/keys/"+ki.KeyName+"/"+ki.KeyVersion, ki.credentials, nil)
```
Comment (gssbzn) [inline internal/decryption/keyproviders/azure.go:88]: [non-blocking] consider using string formatting here instead of concatenation

**Review by gssbzn — APPROVED** [review]
LGTM comments are not blocking and can be a follow up

File: internal/decryption/encrypted_audit_log.go (hunk 2 above)
Comment (fmenezes) [reply internal/decryption/encrypted_audit_log.go:30]: I can remove `omitempty` but we are not generating JSON only reading, so I don't think there will be an effect

Comment (gssbzn) [reply internal/decryption/encrypted_audit_log.go:30]: if this is for deserialization only we are good, I usually just wary of `omitempty` and strings/int/bools as it tends to come back and bite us when serializing

**Review by andreaangiolillo — APPROVED** [review]
LGTM

**Comment by fmenezes** [comment]
@gssbzn the followup ticket is CLOUDP-118858

**Pull request merged**